engine: remove illegal optimization in Seek

tbg opened the pull request:
`Seek` was returning the key-value pair last seeked to when asked
for the same key, but there is no guarantee that the underlying
engine still holds that same key-value pair.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6978)

<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/engine/engine_test.go, line 165 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oFF8AiSrtrdlne9B:-KJ7oFF8AiSrtrdlne9C:1859829904) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L165)):*
> ```Go
> 			}
> 
> 			// Seeks the iterator to `key` and clears the value (but without
> ```

this tense is different than the others =/

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7oJNkpiBeZVeVp2__:-225001819) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
> ```Go
> 			// absent) key.
> 			// The underlying iterator will already be in the right position
> 			// due to its usage in MVCCDelete, and if it reports its cached
> ```

elaborate on this - it is due to its usage in MVCCDelete's retrieval of the meta, not any of the clearing.

---

*[storage/engine/rocksdb.go, line 736 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7nsOtJ3-C8HVLd7rL:-KJ7nsOtJ3-C8HVLd7rM:-425228720) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/rocksdb.go#L736)):*
> ```Go
> 		r.setState(C.DBIterSeekToFirst(r.iter))
> 	} else {
> 		// Its tempting to avoid seeking if we're already at the desired key,
> ```

s/Its/It's/

---

*[storage/engine/rocksdb.go, line 737 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7nvhRfREex_bp7MRu:-KJ7nvhRfREex_bp7MRv:-1303730888) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/rocksdb.go#L737)):*
> ```Go
> 	} else {
> 		// Its tempting to avoid seeking if we're already at the desired key,
> 		// but it may already have changed on the underlying engine; so we
> ```

s/;/,/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978)*
<!-- Sent from Reviewable.io -->

tbg committed:
engine: remove illegal optimization in Seek

`Seek` was returning the key-value pair last seeked to when asked
for the same key, but there is no guarantee that the underlying
engine still holds that same key-value pair.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/engine/engine_test.go, line 165 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oFF8AiSrtrdlne9B:-KJ7qyKcKDFsp8pcOkD6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L165)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this tense is different than the others =/
</details>
Done.

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7qxocjqyg3Slabm7p:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> elaborate on this - it is due to its usage in MVCCDelete's retrieval of the meta, not any of the clearing.
</details>
Done.

---

*[storage/engine/rocksdb.go, line 736 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7nsOtJ3-C8HVLd7rL:-KJ7r2RQK37WWOVWTH92:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/rocksdb.go#L736)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Its/It's/
</details>
Done.

---

*[storage/engine/rocksdb.go, line 737 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7nvhRfREex_bp7MRu:-KJ7r5wd59Q09VxbbUFH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/rocksdb.go#L737)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/;/,/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> #### engine: remove illegal optimization in Seek
> 
> `Seek` was returning the key-value pair last seeked to when asked
> 
> for the same key, but there is no guarantee that the underlying
> 
> engine still holds that same key-value pair.
> 
> 
> 
</details>


---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7sMlJ_hDAy7iy-NDM:-462747781) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
Hm, actually, I bet the Seek in MVCCDelete doesn't matter at all - the only thing that matters is that there's a Clear without informing the iterator. The iterator is already parked there after the MVCCPut.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-:-KJ7s_T8VI6TjjK9_yfu:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7tmznhA3KsyaPyxTJ:712708111) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Hm, actually, I bet the Seek in MVCCDelete doesn't matter at all - the only thing that matters is that there's a Clear without informing the iterator. The iterator is already parked there after the MVCCPut.
</details>
No, when the Put seeks there isn't a value yet (or am I missing a seek?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7u9gi2DWlKJpAlABy:-1017786958) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No, when the Put seeks there isn't a value yet (or am I missing a seek?)
</details>
Oh, good point. So there's even a bug here where if you Put and then Get you won't see the value?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
No, because after the Put the iterator isn't in the right place (same as
before).

On Tue, May 31, 2016 at 6:22 PM Tamir Duberstein <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 1 unresolved
> discussion, some commit checks pending.
> ------------------------------
>
> *storage/engine/engine_test.go, line 175 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ7u9gi2DWlKJpAlABy:-1017786958>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> No, when the Put seeks there isn't a value yet (or am I missing a seek?)
>
>
> Oh, good point. So there's even a bug here where if you Put and then Get
> you won't see the value?
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/6978>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6978#issuecomment-222839287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135NYS-2siK43vrlRS5dyORoC2S79Cks5qHLSpgaJpZM4Iq-Uc>
> .
>
-- 

-- Tobias

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/engine/engine_test.go, line 175 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978#-KJ7oJNkpiBeZVeVp2_Z:-KJ8YomJFx6KE96ewb6E:-212895437) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc4c3276bd275d25d1502779982064375745481e/storage/engine/engine_test.go#L175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Oh, good point. So there's even a bug here where if you Put and then Get you won't see the value?
</details>
No, because after the Put the iterator isn't in the right place (same as before).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6978)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM. We could also fix this by having mutations clear the cached seek data in batch iterations.

petermattis commented:
@tschottdorf Do you mind adding either a `TODO(peter)` or filing an issue regarding my previous comment?

tschottdorf commented:
Sure, I'll file an issue (since I just merged).

On Tue, May 31, 2016 at 9:28 PM Peter Mattis <notifications@github.com>
wrote:

> @tschottdorf <https://github.com/tschottdorf> Do you mind adding either a
> TODO(peter) or filing an issue regarding my previous comment?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6978#issuecomment-222868538>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135CF7DHbLatbI-7LqGPkDwP8PKlubks5qHOAmgaJpZM4Iq-Uc>
> .
>
-- 

-- Tobias

tschottdorf commented:
https://github.com/cockroachdb/cockroach/issues/6980

On Tue, May 31, 2016 at 9:33 PM Tobias Schottdorf <tobias@cockroachlabs.com>
wrote:

> Sure, I'll file an issue (since I just merged).
>
> On Tue, May 31, 2016 at 9:28 PM Peter Mattis <notifications@github.com>
> wrote:
>
>> @tschottdorf <https://github.com/tschottdorf> Do you mind adding either
>> a TODO(peter) or filing an issue regarding my previous comment?
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/6978#issuecomment-222868538>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AE135CF7DHbLatbI-7LqGPkDwP8PKlubks5qHOAmgaJpZM4Iq-Uc>
>> .
>>
> --
>
> -- Tobias
>
-- 

-- Tobias
